module-system/module-implementations/sov-attester-incentives/src/call.rs

```diff
@@ -0,0 +1,764 @@
+            .is_none()
+        {
+            return Err(AttesterIncentiveErrors::UserNotBonded);
+        }
+
+        // We first need to check the bonding proof (because we can't slash the attester if he's not bonded)
+        self.check_bonding_proof(context, &attestation, working_set)?;
+
+        // We suppose that these values are always defined, otherwise we panic
+        let last_attested_height = self
+            .maximum_attested_height
+            .get(working_set)
+            .expect("The maximum attested height should be set at genesis");
+        let current_finalized_height = self
+            .light_client_finalized_height
+            .get(working_set)
+            .expect("The light client finalized height should be set at genesis");
+        let finality = self
+            .rollup_finality_period
+            .get(working_set)
+            .expect("The rollup finality period should be set at genesis");
+
+        // Update the max_attested_height in case the blocks have already been finalized
+        let new_height_to_attest = max(last_attested_height, current_finalized_height) + 1;
+
+        // Minimum height at which the proof of bond can be valid
+        let min_height = if new_height_to_attest > finality {
+            new_height_to_attest - finality
+        } else {
+            0
+        };
+
+        // We have to check the following order invariant is respected:
+        // (height to attest - finality) <= bonding_proof.transition_num <= height to attest
+        //
+        // Which with our variable gives:
+        // min_height <= bonding_proof.transition_num <= new_height_to_attest
+        // If this invariant is respected, we can be sure that the attester was bonded at new_height_to_attest.
+        if !(min_height <= attestation.proof_of_bond.transition_num
+            && attestation.proof_of_bond.transition_num <= new_height_to_attest)
+        {
+            return Err(AttesterIncentiveErrors::InvalidTransitionInvariant);
+        }
+
+        // First compare the initial hashes
+        self.check_initial_hash(
+            new_height_to_attest,
+            context.sender(),
+            &attestation,
+            working_set,
+        )?;
+
+        // Then compare the transition
+        self.check_transition(
+            new_height_to_attest,
+            context.sender(),
+            &attestation,
+            working_set,
+        )?;
+
+        working_set.add_event(
+            "processed_valid_attestation",
+            &format!("attester: {:?}", context.sender()),
+        );
+
+        // Update the maximum attested height
+        self.maximum_attested_height
+            .set(&(new_height_to_attest), working_set);
+
+        // Reward the sender
+        self.reward_sender(
+            context,
+            self.minimum_attester_bond
+                .get(working_set)
+                .expect("Should be defined at genesis"),
+            working_set,
+        )?;
```

Comment:
I think we need to add a conditional to differentiate between the first attestation for a new transition and redundant attestations. 

If the sender is bonded and the attestation is in the right window (from `last_finalized_height` to `current_height`), we should *always* check the attestation and slash the attester if it lies about the transition. 
- When I say "Check the attestation", I mean checking that the da_block_hash and pre_state_root, and post_state_root are correct *for `proof.claimed_transition_num`*.

If the attestation is redundant, we should just stop after that check (we don't slash, but we also don't reward). 
Otherwise, the attestation is the first one for the given transition, so the sender should be rewarded. 

Does that sound right?

Reply:
Sounds right to me, and that's a small fix!